sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp

```diff
@@ -43,73 +43,24 @@ struct joint_matrix {
   }
 };
 
-#ifdef __SYCL_DEVICE_ONLY__
-template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
-          layout Layout>
-class wi_data {
-
-  joint_matrix<Group, T, Use, Rows, Cols, Layout> &jm;
-
-  wi_data(joint_matrix<Group, T, Use, Rows, Cols, Layout> &_jm) : jm(_jm){};
-
-  template <typename Grp, typename Type, use UseJm, size_t NumRows,
-            size_t NumCols, layout LayoutJm>
-  friend decltype(auto)
-  get_wi_data(Grp,
-              joint_matrix<Grp, Type, UseJm, NumRows, NumCols, LayoutJm> &);
-
-public:
-  size_t length() {
-#if defined(__NVPTX__)
-    return jm.cuda_impl.wi_marray.size();
-#else
-    return __spirv_JointMatrixWorkItemLengthINTEL(jm.spvm);
-#endif
-  };
-
-  decltype(auto) operator[](size_t i) {
+template <typename Group, typename T, use Use, size_t M, size_t N,
+          layout Layout, typename F>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &C,
+                   F &&lambda) {
+#if defined(__SYCL_DEVICE_ONLY__)
 #if defined(__NVPTX__)
-    return (jm.cuda_impl.wi_marray[i]);
-#else
-    return wi_element<T, Rows, Cols, Use, Layout, Group>(jm, i);
+#else // NVPTX
```

Comment:
@JackAKirk, what is the plan on adding CUDA specific implementation? Can we have a CUDA PR that adds implementation for this so we can merge them around the same time?

Reply:
I've suggested the cuda implementation in a comment above. It is simple and can be added here if you like.

I think the best way forward is 
- Keep the nvidia implementation of wi_data/jm.get_wi_data() but deprecate them (if you want to move wi_data to intel namespace that is fine, nvidia implementation of wi_data can live in that namespace whilst it is deprecated because it is not constructable anyway and this won't break any code). I don't think you have to touch the nvidia wi_data/get_wi_data code at all: you can just move them to the intel header as you've done and intel namespace if you desire this.
- Add the simple nvidia implementation I suggested in this PR for `joint_matrix_apply`.

Then at a later time we can simply remove the deprecated implementations of wi_data and get_wi_data for cuda only.

Reply:
@JackAKirk Please take a look if you are OK with this PR.

Reply:
I'll update a cuda test shortly to test everything is fine.

Reply:
Resolving as @JackAKirk confirmed test is working fine.